pkg/os/volume/api.go

```diff
@@ -268,6 +277,96 @@ func getTarget(mount string) (string, error) {
 	return volumeString, nil
 }
 
+// GetVolumeIDFromTargetPath returns the volume id of a given target path.
+func (VolumeAPI) GetClosestVolumeIDFromTargetPath(targetPath string) (string, error) {
+	volumeString, err := findClosestVolume(targetPath)
+
+	if err != nil {
+		return "", fmt.Errorf("error getting the closest volume for the path=%s, err=%v", targetPath, err)
+	}
+
+	return volumeString, nil
+}
+
+// findClosestVolume finds the closest volume id for a given target path
+// by following symlinks and moving up in the filesystem, if after moving up in the filesystem
+// we get to a DriveLetter then the volume corresponding to this drive letter is returned instead.
+func findClosestVolume(path string) (string, error) {
+	candidatePath := path
+
+	// run in a bounded loop to avoid doing an infinite loop
+	// while trying to follow symlinks
+	for i := 0; i < 256; i += 1 {
+		fi, err := os.Lstat(candidatePath)
+		if err != nil {
+			return "", err
+		}
+		isSymlink := fi.Mode()&os.ModeSymlink != 0
+
+		if isSymlink {
+			target, err := dereferenceSymlink(candidatePath)
+			if err != nil {
+				return "", err
+			}
+			// if it has the form Volume{volumeid}\ then it's a volume
+			if VolumeRegexp.Match([]byte(target)) {
+				// symlinks that are pointing to Volumes don't have this prefix
+				target = "\\\\?\\" + target
+				return target, nil
+			}
+			// otherwise follow the symlink
+			candidatePath = target
+		} else {
+			// if it's not a symlink move one level up
+			previousPath := candidatePath
+			candidatePath = filepath.Dir(candidatePath)
+
+			// if the new path is the same as the previous path then we reached the root path
+			if previousPath == candidatePath {
+				// find the volume for the root path (assuming that it's a DriveLetter)
+				target, err := getVolumeForDriveLetter(candidatePath[0:1])
+				if err != nil {
+					return "", err
+				}
+				return target, nil
+			}
+		}
+
+	}
+
+	return "", fmt.Errorf("Failed to find the closest volume for path=%s", path)
+}
+
+// dereferenceSymlink dereferences the symlink `path` and returns the stdout.
+func dereferenceSymlink(path string) (string, error) {
+	cmd := exec.Command("powershell", "/c", fmt.Sprintf(`(Get-Item -Path %s).Target`, path))
+	klog.V(8).Infof("About to execute: %q", cmd.String())
+	targetb, err := cmd.Output()
```

Comment:
not sure it is the case, sometimes the error is empty, but command actually return some error messages, so targetb will be some strings. In this case just want to make sure the logic can handle it.

Reply:
true, `cmd.Output` returns the combined output, places where this is called ensure that the path sent is a symlink however another way to test it would be by analyzing stderr apart from stdout.

The existing `GetVolumeIDFromTargetPath` doesn't do a separate analysis either, it's using the combined output and relying on err for errors

Reply:
I just wondering whether it is possible to make sure it return a valid path or a volume id, instead of some long string

Reply:
I think that if the target doesn't exist it'll return an error that's propagated, so at this point we know that it's a path or a volumeId, for anything in stderr I'm not checking though, I'll only analyze the output of stdout and error if stderr has contents